docs/design/graceful-domain-failover/3051-graceful-domain-failover.md

```diff
@@ -0,0 +1,230 @@
+
+## Implementation
+
+New components:
+1. New state in domain
+2. New task processor
+3. Failover marker
+4. Failover coordinator
+5. Buffer queue
+
+### Domain
+
+A new state "Pending_Active "introduced when domain moves from Passive to Active.
+![Domain state transition](3051-state-transition.png)
+
+Active to Passive: This happens when a domain failover from ‘active’ to ‘passive’ in the cluster.
+
+Passive to Pending_Active: This happens when a domain failover from ‘passive’ to ‘active’ in the cluster. In this pending-active cluster, it first updates domain state to pending.
+
+Pending_Active to Passive: This happens when the domain is in ‘pending_active’, the coordinator receives a domain failover notification with a higher version and the domain failovers to another cluster. Then the domain moves back to passive.
+
+Pending_Active to Active: The coordinator moves domain from ‘pending_active’ to ‘active’ in the scenarios:
+1. All shards received the failover notification and failover markers. 
+2. The failover timeout reaches and the domain is not ‘active’.
+
+### Task processor
+
+As the new state introduced during graceful domain failover in the passive cluster, new task processor introduce here to handle the task in Pending_Active state.
+
+Transfer: No ops during failover.
+Timer: Blocked on processing task during failover.
+
+### Failover marker
+FailoverMarker {
+
+    *replicationTask
+    failoverVersion int64
+    sourceCluster string
+    Timestamp time.Time
+}
+
+### Failover coordinator
+With the graceful failover protocol, we need to maintain a global state (in the same cluster) of all shards. So we need a new component for it. The coordinator could be a stand-alone component or elect a leader from the shards. This new component is to maintain a global state of all shards during a failover.
+To maintain the global state
+Each shard does heartbeat to the coordinator to send the last X minutes failover marker (X is the max graceful failover timeout we support).
+
+The coordinator persists the state in memory and updates this state to database periodically. The state can be stored in the shard table. The state struct looks like:
+map[string][]*int32
+The key contains the domain and the failover version.
+The value is a slice of shard ID.
+Failover timeout
+Currently, each history host has a component domain cache. Each shard on the same host gets domain failover notification from the domain cache. Domain cache periodically checks the database and updates the domain in memory. The failover timeout can leverage this component.
+
+During graceful failover domain update, we record the timeout in the domain data. Domain cache reads all domain data periodically and checks if any of the graceful failover should be timed out. If the domain cache get a graceful failover to be timed out. It can sends a notification to shard to update the domain from pending_active to active.
+
+### Buffer queue
```

Comment:
I would advise against adding this feature. The reason there is no buffering of these calls as they provide consistency guarantees. For example Signal should fail if workflow is already closed. 
I would reject requests during failover and rely on clients to retry. 
If someone really wants to skip on consistency they can always buffer requests in Kafka.

Reply:
Good suggestion. I think this is not the MVP and the feature can be configurable by domains. However, it depends on the user's requirements. We might still think about this feature.

In terms of consistency, it calls to the source cluster to get workflow state before buffering the request. If the workflow is closed, then Signal will fail.

Reply:
Agreed that buffer queue isn't part of the core requirements at this stage. Leaving it out also simplifies the design and creates crisp boundaries.

On the other hand, my understanding is that there are customers who want to have graceful failovers without sacrificing the availability of the synchronous APIs (Start, Signal, etc.). Asking them to create their own buffer is a viable option, but this will probably become a common enough pattern that it might make sense to solve at our end eventually.  Besides, if we do it for them, we can do more than merely buffering the requests in a simple queue such as checking the workflow status from the source cluster as @yux0 mentioned or providing some more guarantees around not buffering conflicting requests such as terminating workflow and then signaling it. 

In a later phase, I believe we should make the destination cluster immediately available for newly started workflows during graceful failovers, such that the cluster simply acts as "Active" for new workflows, rather than acting like "PendingActive". That would limit the impact of downtime and/or latency caused by graceful failover only to the workflows that were in-flight before the failover was triggered. This "graceful failover with immediate availability for new workflows" enhancement is currently just an idea that we haven't explored yet, but I think the availability and latency gains will be good enough to justify exploring it. If we do that, some sort of request buffering will make even more sense in order to provide a consistent experience for all in-flight workflows regardless of when exactly they were started.

Let's leave out request buffering in this phase, but keep it on the table for the next iteration. In the meantime, I'd love to hear your thoughts on the points I mentioned above.

Reply:
I think the problem with buffering is that you have to deal with invalid buffered requests. Then you start inventing DLQs and similar concepts. 
I believe this is premature optimization. Under normal situation graceful failover should be under 3 seconds, well within latency spike due to all sort of other types of failures.